packages/components/src/slot-fill/types.ts

```diff
@@ -0,0 +1,119 @@
+/**
+ * External dependencies
+ */
+import type { Component, MutableRefObject, ReactNode } from 'react';
+
+export type SlotKey = string | symbol;
+
+type SlotPropBase = {
+	/**
+	 * Slot name.
+	 */
+	name: SlotKey;
+	/**
+	 * props to pass from `Slot` to `Fill`.
+	 *
+	 * @default {}
+	 */
+	fillProps?: any;
+
+	/**
+	 * By default, events will bubble to their parents on the DOM hierarchy (native event bubbling).
+	 * If set to true, events will bubble to their virtual parent in the React elements hierarchy instead,
+	 * also accept an optional `className`, `id`, etc.  to add to the slot container.
+	 */
+	bubblesVirtually?: boolean;
+};
+
+export type SlotComponentProps =
+	| ( SlotPropBase & {
+			bubblesVirtually: true;
+	  } )
+	| ( SlotPropBase & {
+			bubblesVirtually?: false;
+
+			/**
+			 * A function that returns nodes to be rendered.
+			 * Not supported when `bubblesVirtually` is true.
+			 *
+			 * @param fills
+			 */
+			children?: ( fills: ReactNode ) => ReactNode;
```

Comment:
Could be an improvement to try with the distributive types patterns and mark `children` as `never` when `bubblesVirtually` is `true`. That pattern is utilized here:

https://github.com/WordPress/gutenberg/blob/fbf1bcd6316235865b456820f63e51026370a4a2/packages/components/src/animate/types.ts#L26

Reply:
```
export type SlotComponentProps< T extends boolean > = SlotPropBase & {
	bubblesVirtually: T;
	children?: T extends true ? never : ( fills: ReactNode ) => ReactNode;
};
```

I have tried the above code and others, but the type inference does not work. :vomiting_face:

Reply:
That's a bummer. It seems to work in some other components - did you spot any difference in how we're using it? Perhaps this is a good opportunity to report a TS bug upstream.

Reply:
Perhaps the type of <Slot /> changes with or without bubblesVirtually, or the internal <Slot /> does not pass bubblesVirtually, which complicates the type definition.

Rather than a bug in TypeScript, I believe it is due to the complexity of the many places where SlotComponentProps are applied.